### PR TITLE
allure-reporter: fixed bug with mocha all hooks being treated as tests

### PR DESCRIPTION
**Pull request by Gennadiii**
## Proposed changes

[//]: # Before and After hooks appear in report like tests which breaks statistics of tests. I might have 1 failing test out of 1 and it's 100% fails but with successful hooks it's only 33%.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ x] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by christian-bromann** [comment]
So hooks are seen in allure as tests? Is there a concept like hooks in Allure?

cc @BorisOsipov

**Comment by Gennadiii** [comment]
Yes, they are. You run 10 tests, 2 fail and you get 6.6% of failed
tests instead of 20%. But my fix is probably good for jasmine but
definitely bad for mocha. I'm talking about it with Boris.
Not sure about Allure but don't think they have "hook" concept.

File: packages/wdio-allure-reporter/src/index.js [hunk 1]
```diff
@@ -199,13 +199,15 @@ class AllureReporter extends WDIOReporter {
             return
         }
 
-        // add hook as test to suite
-        this.onTestStart(hook)
+        // add hook as test to suite if not mocha All hook
+        if (!isMochaAllHooks(hook.title)) {
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.js]: To be consistent with other if's
```
if (isMochaAllHooks(hook.title)) {
  return
}

this.onTestStart(hook);
```

File: packages/wdio-allure-reporter/src/index.js [hunk 2]
```diff
@@ -221,8 +223,11 @@ class AllureReporter extends WDIOReporter {
 
         // set hook (test) status
         if (hook.error) {
-            this.onTestFail(hook)
-        } else {
+            if (isMochaAllHooks(hook.title)) {
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/src/index.js]: is it necessary to start test here? See https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-allure-reporter/src/index.js#L123

File: packages/wdio-allure-reporter/src/index.js (hunk 1 above)
Comment (Gennadiii) [reply packages/wdio-allure-reporter/src/index.js]: BTW do we even need to add hook as test to suite on the last line? We ignore global hooks, have handles for each and all hooks. Which hooks are left?

Comment (Gennadiii) [reply packages/wdio-allure-reporter/src/index.js]: > To be consistent with other if's
> 
> ```
> if (isMochaAllHooks(hook.title)) {
>   return
> }
> 
> this.onTestStart(hook);
> ```

BTW do we even need to add hook as test to suite on the last line? We ignore global hooks, have handles for each and all hooks. Which hooks are left?

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.js]: e.g. some cucumber hooks. I am not sure how reporter handles them

File: packages/wdio-allure-reporter/src/index.js (hunk 2 above)
Comment (Gennadiii) [reply packages/wdio-allure-reporter/src/index.js]: In my opinion it's more readable. If I remove it then there's onTest Fail or Pass without onTestStart. So we finishing without starting. Thou if you insist I'll remove it, np.

Comment (BorisOsipov) [reply packages/wdio-allure-reporter/src/index.js]: ok makes sense

File: packages/wdio-allure-reporter/tests/allureFeatures.test.js [hunk 3]
```diff
@@ -462,6 +461,21 @@ describe('hooks handling', () => {
...
+    it('should ignore mocha all hooks if hook passes', () => {
+        reporter.allure = allureInstance()
+        reporter.onHookEnd({ title: '"before all" hook', parent: 'foo' })
+
+        expect(endStep).toHaveBeenCalledTimes(0)
+    })
+
+    it('should treat mocha all hooks as tests if hook throws', () => {
+        reporter.allure = allureInstance()
+        reporter.onHookEnd({ title: '"before each" hook', parent: 'foo', error: { message: '', stack: '' } })
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/tests/allureFeatures.test.js]: it isn't `mocha all hook`

File: packages/wdio-allure-reporter/tests/allureFeatures.test.js [hunk 4]
```diff
@@ -462,6 +461,21 @@ describe('hooks handling', () => {
...
+
+        expect(endStep).toHaveBeenCalledTimes(1)
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/tests/allureFeatures.test.js]: add expect(onTestStart) and expect(onTestFail) called

File: packages/wdio-allure-reporter/tests/allureFeatures.test.js [hunk 5]
```diff
@@ -462,6 +461,21 @@ describe('hooks handling', () => {
         expect(endStep).toHaveBeenCalledTimes(1)
         expect(endStep.mock.results[0].value).toBe('failed')
     })
+
+    it('should ignore mocha all hooks if hook passes', () => {
+        reporter.allure = allureInstance()
+        reporter.onHookEnd({ title: '"before all" hook', parent: 'foo' })
+
+        expect(endStep).toHaveBeenCalledTimes(0)
```
Comment (BorisOsipov) [inline packages/wdio-allure-reporter/tests/allureFeatures.test.js]: 1) not `endStep` but `endCase`
2) also please add test that `reporter.onHookStart({ title: '"before all" hook', parent: 'foo' })` startCase isnt invoked

File: packages/wdio-allure-reporter/tests/allureFeatures.test.js (hunk 3 above)
Comment (Gennadiii) [reply packages/wdio-allure-reporter/tests/allureFeatures.test.js]: Thanks, missed that

File: packages/wdio-allure-reporter/tests/allureFeatures.test.js (hunk 4 above)
Comment (Gennadiii) [reply packages/wdio-allure-reporter/tests/allureFeatures.test.js]: How do I do that? Don't see such examples

**Comment by christian-bromann** [comment]
@Gennadiii can you please sign the CLA?

**Comment by Gennadiii** [comment]
> @Gennadiii can you please sign the CLA?

I already did, didn't I? Please recheck.

**Comment by christian-bromann** [comment]
> I already did, didn't I? Please recheck.

Nope, license

**Comment by christian-bromann** [comment]
Make sure you sign with the same email that you use for commits

**Comment by Gennadiii** [comment]
> > I already did, didn't I? Please recheck.
> 
> Nope, license still not signed: `licence/cla Pending — Contributor License Agreement is not signed yet.`

Yeah, I see that pending status but when I go to details it's already filled with my data from this account and checkbox is checked and I can't do any changes

**Comment by christian-bromann** [comment]
> Yeah, I see that pending status but when I go to details it's already filled with my data from this account and checkbox is checked and I can't do any changes

Make sure you sign with the same email that you used to make the commit!

**Comment by Gennadiii** [comment]
> > Yeah, I see that pending status but when I go to details it's already filled with my data from this account and checkbox is checked and I can't do any changes
> 
> Make sure you sign with the same email that you used to make the commit!

I have a single github account

Screenshot_1
Screenshot_2

**Comment by christian-bromann** [comment]
Your commit was not made with that email:

```
commit 9fd79b87e27695e125e194848516a8bfe6383c09 (HEAD, Gennadiii/master)
Author: Gennadii_Mishchevskii <gmishchevskii@cogniance.com>
Date:   Tue Aug 27 17:21:50 2019 +0300
```

**Comment by Gennadiii** [comment]
> Your commit was not made with that email:
> 
> ```
> commit 9fd79b87e27695e125e194848516a8bfe6383c09 (HEAD, Gennadiii/master)
> Author: Gennadii_Mishchevskii <gmishchevskii@cogniance.com>
> Date:   Tue Aug 27 17:21:50 2019 +0300
> ```

That's unexpected and not clear... I don't have that email for over 2 years.

**Comment by christian-bromann** [comment]
> I don't have that email for over 2 years.

This is how you have set up git. You can change the email and amend the commit so that the email is updated.

**Comment by Gennadiii** [comment]
> > I don't have that email for over 2 years.
> 
> This is how you have set up git. You can change the email and amend the commit so that the email is updated.

Did that, thanks. It was confusing - old local settings

**Pull request merged**